libdeno/binding.cc

```diff
@@ -208,17 +210,35 @@ void Print(const v8::FunctionCallbackInfo<v8::Value>& args) {
   stream << cstr << std::endl;
 }
 
-static v8::Local<v8::Uint8Array> ImportBuf(v8::Isolate* isolate, deno_buf buf) {
+static v8::Local<v8::Uint8Array> ImportBuf(Deno* d, deno_buf buf) {
   if (buf.alloc_ptr == nullptr) {
     // If alloc_ptr isn't set, we memcpy.
     // This is currently used for flatbuffers created in Rust.
-    auto ab = v8::ArrayBuffer::New(isolate, buf.data_len);
+
+    // To avoid excessively allocating new ArrayBuffers, we try to reuse a
+    // single global ArrayBuffer. The caveat is that users must extract data
+    // from it before the next tick. We only do this for ArrayBuffers less than
+    // 1024 bytes.
+    v8::Local<v8::ArrayBuffer> ab;
+    if (buf.data_len > GLOBAL_IMPORT_BUF_SIZE) {
+      // Simple case. We allocate a new ArrayBuffer for this.
+      ab = v8::ArrayBuffer::New(d->isolate, buf.data_len);
+    } else {
+      // Fast case. We reuse the global ArrayBuffer.
+      if (d->global_import_buf.IsEmpty()) {
+        // Lazily initialize it.
+        ab = v8::ArrayBuffer::New(d->isolate, GLOBAL_IMPORT_BUF_SIZE);
+        d->global_import_buf.Reset(d->isolate, ab);
+      } else {
+        ab = d->global_import_buf.Get(d->isolate);
+      }
+    }
     memcpy(ab->GetContents().Data(), buf.data_ptr, buf.data_len);
     auto view = v8::Uint8Array::New(ab, 0, buf.data_len);
```

Comment:
This is also an expensive call. Ideally we'd be able to reuse the views if we reuse the ArrayBuffer.